File: tests/testthat/test-contract.R
# Copyright 2026 Google LLC

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     https://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

test_that("longbet() formals match the shared API contract", {
  skip_if_not_installed("yaml")
  path <- system.file("contract", "longbet-api.yaml", package = "longbet")
  if (!nzchar(path)) path <- file.path("..", "..", "contract", "longbet-api.yaml")
  skip_if_not(file.exists(path), "contract file not found")

  contract <- yaml::read_yaml(path)
  formals_lb <- formals(longbet)

  for (arg in contract$arguments) {
    if (!isTRUE(arg$expose_in_r)) next
    r_name <- arg$r_name
    expect_true(r_name %in% names(formals_lb),
                info = sprintf("longbet() is missing the contract argument '%s'", r_name))

    default <- formals_lb[[r_name]]
    if (is.call(default) && identical(as.character(default[[1]]), "c")) {
      # match.arg style: the first choice is the default
      default <- eval(default)[[1]]
    } else if (is.name(default)) {
      next
    } else {
      default <- eval(default)
    }
    expected <- arg$default
    if (is.null(expected)) {
      expect_null(default, info = sprintf("default for '%s'", r_name))
    } else if (is.numeric(expected) && is.numeric(default)) {
      expect_equal(as.numeric(default), as.numeric(expected), tolerance = 1e-12,
                   info = sprintf("default for '%s'", r_name))
    } else {
      expect_equal(default, expected, info = sprintf("default for '%s'", r_name))
    }
  }
})

test_that("every exposed contract argument really reaches the engine", {
  skip_without_engine()
  skip_if_not_installed("yaml")
  path <- system.file("contract", "longbet-api.yaml", package = "longbet")
  if (!nzchar(path)) path <- file.path("..", "..", "contract", "longbet-api.yaml")
  skip_if_not(file.exists(path), "contract file not found")

  contract <- yaml::read_yaml(path)
  lb <- reticulate::import("longbet")
  cfg <- lb$LongBetConfig()
  py_defaults <- reticulate::py_to_r(cfg$to_dict())

  for (arg in contract$arguments) {
    py_name <- arg$python_name
    expect_true(py_name %in% names(py_defaults),
                info = sprintf("LongBetConfig has no field '%s'", py_name))
  }
})

File: tests/test_contract.py
# Copyright 2026 Google LLC

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     https://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""The Python side of the two-front-door contract (contract/longbet-api.yaml).

Checked in **both** directions: every contract entry must match the dataclass,
and every dataclass field must appear in the contract or be declared internal.
A one-directional check cannot catch a new option that was never written down.
"""

import dataclasses
from pathlib import Path

import pytest
import yaml

from longbet import LongBetConfig

CONTRACT_PATH = Path(__file__).resolve().parents[1] / "contract" / "longbet-api.yaml"


@pytest.fixture(scope="module")
def contract():
    assert CONTRACT_PATH.exists(), f"contract file not found: {CONTRACT_PATH}"
    with open(CONTRACT_PATH, encoding="utf-8") as f:
        return yaml.safe_load(f)


def _defaults():
    return {f.name: f.default for f in dataclasses.fields(LongBetConfig)}


def test_every_contract_argument_matches_the_config(contract):
    defaults = _defaults()
    for arg in contract["arguments"]:
        name = arg["python_name"]
        assert name in defaults, f"contract lists {name!r}, which LongBetConfig lacks"
        assert defaults[name] == arg["default"], (
            f"default mismatch for {name}: contract says {arg['default']!r}, "
            f"LongBetConfig says {defaults[name]!r}"
        )


def test_every_config_field_is_in_the_contract(contract):
    documented = {a["python_name"] for a in contract["arguments"]}
    internal = set(contract.get("internal", []))
    undocumented = set(_defaults()) - documented - internal
    assert not undocumented, (
        f"LongBetConfig fields missing from the contract: {sorted(undocumented)}. "
        "Add them to contract/longbet-api.yaml (and to the R front door), or list "
        "them under `internal`."
    )


def test_r_names_are_unique(contract):
    r_names = [a["r_name"] for a in contract["arguments"] if a.get("expose_in_r")]
    assert len(r_names) == len(set(r_names)), "duplicate r_name in the contract"


def test_yaml_files_are_byte_identical():
    c1 = (Path(__file__).resolve().parents[1] / "contract" / "longbet-api.yaml").read_bytes()
    c2 = (Path(__file__).resolve().parents[1] / "inst" / "contract" / "longbet-api.yaml").read_bytes()
    assert c1 == c2, "contract/longbet-api.yaml and inst/contract/longbet-api.yaml differ"


def test_multi_contract_structure(contract):
    assert "multi_api" in contract, "contract missing multi_api section"
    m_api = contract["multi_api"]
    from longbet._sur import SAMPLER_SEMANTICS
    from longbet._io import PRECISION_CACHE_VERSION
    assert m_api.get("sampler_semantics") == SAMPLER_SEMANTICS
    assert m_api.get("precision_cache_version") == PRECISION_CACHE_VERSION
    assert "explicitly positive sigma_prior_a and sigma_prior_b" in m_api["variance_prior_requirement"]
    assert "inputs" in m_api
    assert "fit_returns" in m_api
    assert "predict_returns" in m_api
    assert "utilities" in m_api
    for util in ("effect_draws", "joint_prob", "outcome_correlation"):
        assert util in m_api["utilities"]


def test_python_multi_exports():
    import longbet
    for symbol in (
        "LongBetMulti",
        "LongBetMultiPrediction",
        "effect_draws",
        "joint_prob",
        "outcome_correlation",
        "effect_draws_from_arrays",
        "reduce_joint_masks",
    ):
        assert hasattr(longbet, symbol), f"longbet missing export {symbol}"
        assert symbol in longbet.__all__, f"{symbol} missing from longbet.__all__"


def test_r_frontdoor_signatures_and_defaults(contract):
    """Ensure R functions longbet and longbet_multi have formals matching the contract."""
    import re
    repo_root = Path(__file__).resolve().parents[1]

    # Parse R/longbet.R
    lb_r = (repo_root / "R" / "longbet.R").read_text(encoding="utf-8")
    lb_sig_match = re.search(r"longbet\s*<-\s*function\s*\((.*?)\)\s*\{", lb_r, re.DOTALL)
    assert lb_sig_match, "Could not find longbet signature in R/longbet.R"
    lb_params = [p.strip().split("=")[0].strip() for p in lb_sig_match.group(1).split(",") if p.strip()]

    # Parse R/multi.R
    multi_r = (repo_root / "R" / "multi.R").read_text(encoding="utf-8")
    multi_sig_match = re.search(r"longbet_multi\s*<-\s*function\s*\((.*?)\)\s*\{", multi_r, re.DOTALL)
    assert multi_sig_match, "Could not find longbet_multi signature in R/multi.R"
    multi_params = [p.strip().split("=")[0].strip() for p in multi_sig_match.group(1).split(",") if p.strip()]

    # Check exposed arguments in both functions
    for arg in contract["arguments"]:
        if not arg.get("expose_in_r"):
            continue
        r_name = arg["r_name"]
        assert r_name in lb_params, f"longbet in R/longbet.R missing formal '{r_name}'"
        assert r_name in multi_params, f"longbet_multi in R/multi.R missing formal '{r_name}'"

    # Check NAMESPACE exports
    ns = (repo_root / "NAMESPACE").read_text(encoding="utf-8")
    for exp in (
        "export(longbet_multi)",
        "export(effect_draws)",
        "export(joint_prob)",
        "export(outcome_correlation)",
        "S3method(predict, longbet_multi)",
        "S3method(print, longbet_multi)",
        "S3method(summary, longbet_multi)",
        "S3method(print, longbet_multi.pred)",
        'S3method("[", longbet_multi)',
        'S3method("[", longbet_multi.pred)',
    ):
        assert exp in ns, f"NAMESPACE missing {exp}"
